File: src/faz/client/go_client/faz_client.go
package faz_client

import (
	"log"
	"os"
	"fmt"

	"github.com/golang/protobuf/ptypes/any"
	"golang.org/x/net/context"
	"google.golang.org/grpc"

	faz "github.com/floan/csci499_floan/protos/go_protos/faz"
)

type FazClient struct {
	stub_ faz.FazServiceClient
	conn_ *grpc.ClientConn
}

// This function acts as a constructor 
// and initializes the member variables 
// of the FazClient struct
func CreateFazClient(address string) *FazClient {
	// Set up a connection to the server.
	fazClient := &FazClient{}
	conn, err := grpc.Dial(address, grpc.WithInsecure())
	fazClient.conn_ = conn
	if err != nil {
		log.Fatalf("did not connect: %v", err)
		os.Exit(1)
	}
	// Creating the stub that will call methods
	fazClient.stub_ = faz.NewFazServiceClient(fazClient.conn_)
	// Returning our newly created object
	return fazClient
}

// This function acts as a destructor 
// and does what is needed to destroy the 
// object. NOTE: THIS HAS TO BE CALLED MANUALLY
func (fazClient *FazClient) DestroyFazClient() {
	fazClient.conn_.Close()
}

// This method links to the event method in the server class
// This executes our stored function via our FaaS Faz
// Args: A request and response of protobuf Any type. These
//       represent the input and output from our registered function
// Returns: A boolean indicating success / failure
func (fazClient *FazClient) Event(request *any.Any, 
																	response *any.Any,
																	eventType int) bool {
	eventRequest := faz.EventRequest{
		EventType: int32(eventType), 
		Payload: request,
	}
	fmt.Println(eventRequest)
	eventResponse, err := fazClient.stub_.Event(context.Background(), &eventRequest)
	if err != nil {
		log.Fatalf("could not perform event: %v", err)
		return false
	} else {
		// Setting the Any payload in eventResponse to
		// our Any response variable
		response = eventResponse.GetPayload()
		fmt.Println(eventResponse)
		return true
	}
}

// This method calls the Faz hook function. This is supposed to
// hook a custom function to our FaaS.
// Args: An event type that corresponds with the function call
//       And a string that represents the function name
// Returns: A boolean indicating success / failure
func (fazClient *FazClient) HookFunction(eventType int,
																				 functionName string) bool {
	hookRequest := faz.HookRequest{
		EventType: int32(eventType),
		EventFunction: functionName,
	}
	_, err := fazClient.stub_.Hook(context.Background(), &hookRequest)
	if err != nil {
		log.Fatalf("could not hook function: %v", err)
		return false
	} else {
		return true
	}
}

// This method calls the faz unhook function. This is supposed
// to unhook a custom function to our FaaS.
// Args: An event type whose corresponding function we are to
//       unhook.
// Returns: A boolean indicating success / failure
func (fazClient *FazClient) UnhookFunction(eventType int) bool {
	unhookRequest := faz.UnhookRequest{
		EventType: int32(eventType),
	}
	_, err := fazClient.stub_.Unhook(context.Background(), &unhookRequest)
	if err != nil {
		log.Fatalf("could not unhook function: %v", err)
		return false
	} else {
		return true
	}
}

File: src/faz/client/cpp_client/faz_client.cpp
#include "faz_client.h"

#include <iostream>
#include <string>
#include <vector>

#include "../../../caw/functions/caw_functions.h"

Status FazClient::Event(Any request, Any *response, int eventType) {
  // Setting up request/response wrappers
  EventRequest eventRequest;
  EventReply eventResponse;
  // Creating grpc context variable
  ClientContext context;
  *(eventRequest.mutable_payload()) = request;
  eventRequest.set_event_type(eventType);

  Status status = stub_->event(&context, eventRequest, &eventResponse);
  *response = *eventResponse.mutable_payload();

  if (!status.ok()) {
    LOG(ERROR) << "Something went wrong";
    LOG(ERROR) << status.error_code() << ": " << status.error_message();
  }

  return status;
}

Status FazClient::StreamEvent(const Any request,
                              std::function<void(EventReply)> &callback,
                              int eventType) {
  // setup vars for sending to server
  EventRequest event_request;
  ClientContext context;
  *(event_request.mutable_payload()) = request;
  event_request.set_event_type(eventType);
  std::unique_ptr<ClientReader<EventReply> > reader(
      stub_->streamEvent(&context, event_request));
  EventReply reply;
  while (reader->Read(&reply)) {
    callback(reply);
  }
  return reader->Finish();
}

Status FazClient::HookFunction(int eventType, std::string functionName) {
  HookRequest request;
  HookReply response;
  ClientContext context;

  request.set_event_type(eventType);
  request.set_event_function(functionName);

  Status status = stub_->hook(&context, request, &response);

  if (!status.ok()) {
    LOG(ERROR) << "Something went wrong";
    LOG(ERROR) << status.error_code() << ": " << status.error_message();
  }

  return status;
}

Status FazClient::UnhookFunction(int eventType) {
  UnhookRequest request;
  UnhookReply response;
  ClientContext context;

  request.set_event_type(eventType);

  Status status = stub_->unhook(&context, request, &response);

  if (!status.ok()) {
    LOG(ERROR) << "Something went wrong";
    LOG(ERROR) << status.error_code() << ": " << status.error_message();
  }

  return status;
}
